lib/hyperreduction/S_OPT.C

```diff
@@ -0,0 +1,531 @@
+    blockcounts[0] = 1;
+
+    /*
+       MPI_Type_extent is deprecated in MPI-2 and removed in
+       MPI-3. MPI_Type_extent has been replaced by MPI_Type_get_extent.
+    */
+#if (MPI_VERSION == 1)
+    MPI_Type_extent(MPI_DOUBLE, &extent);
+#else
+    MPI_Type_get_extent(MPI_DOUBLE, &lower_bound, &extent);
+#endif
+
+    offsets[1] = extent;
+    oldtypes[1] = MPI_INT;
+    blockcounts[1] = 2;
+
+    /*
+      MPI_Type_struct is deprecated in MPI-2 and removed in
+      MPI-3. MPI_Type_struct has been replaced by MPI_Type_create_struct.
+     */
+#if (MPI_VERSION == 1)
+    MPI_Type_struct(2, blockcounts, offsets, oldtypes, &MaxRowType);
+#else
+    MPI_Type_create_struct(2, blockcounts, offsets, oldtypes, &MaxRowType);
+#endif
+    MPI_Type_commit(&MaxRowType);
+
+    // Create an MPI_Op for the RowInfoMax function.
+    MPI_Op RowInfoOp;
+    MPI_Op_create((MPI_User_function*)RowInfoMax, true, &RowInfoOp);
+
+    // Get the number of basis vectors and the size of each basis vector.
+    CAROM_VERIFY(0 < num_f_basis_vectors_used && num_f_basis_vectors_used <= f_basis->numColumns());
+    const int num_basis_vectors =
+        std::min(num_f_basis_vectors_used, f_basis->numColumns());
+    const int num_samples = num_samples_req > 0 ? num_samples_req : num_basis_vectors;
+    CAROM_VERIFY(num_basis_vectors <= num_samples && num_samples <= f_basis->numDistributedRows());
+    CAROM_VERIFY(num_samples == f_sampled_row.size());
+    CAROM_VERIFY(num_samples == f_basis_sampled_inv.numRows() &&
+                 num_basis_vectors == f_basis_sampled_inv.numColumns());
+    CAROM_VERIFY(!f_basis_sampled_inv.distributed());
+
+    int num_rows = f_basis->numRows();
+
+    // If num_basis_vectors is less than the number of columns of the basis,
+    // we need to truncate the basis.
+    const Matrix* f_basis_truncated = NULL;
+    if (num_basis_vectors < f_basis->numColumns())
+    {
+        f_basis_truncated = f_basis->getFirstNColumns(num_basis_vectors);
+    }
+    else
+    {
+        f_basis_truncated = f_basis;
+    }
+
+    const Matrix* Vo = NULL;
+
+    std::vector<int> row_offset(num_procs + 1);
+    row_offset[num_procs] = f_basis_truncated->numDistributedRows();
+    row_offset[myid] = num_rows;
+
+    CAROM_VERIFY(MPI_Allgather(MPI_IN_PLACE,
+                               1,
+                               MPI_INT,
+                               row_offset.data(),
+                               1,
+                               MPI_INT,
+                               MPI_COMM_WORLD) == MPI_SUCCESS);
+    for (int i = num_procs - 1; i >= 0; i--) {
+        row_offset[i] = row_offset[i + 1] - row_offset[i];
+    }
+
+    CAROM_VERIFY(row_offset[0] == 0);
+
+    SLPK_Matrix slpk_f_basis;
+
+    int nrow_blocks = num_procs;
+    int ncol_blocks = 1;
+
+    // Use the QR factorization of the input matrix, if requested
+    if (qr_factorize)
+    {
+        int blocksize = row_offset[num_procs] / num_procs;
```

Comment:
Could this block of code be a member function in `Matrix` to return a `Matrix*` for the QR factorization?